### PR TITLE
[READY] Multi-Z Openspace visual fixes.

### LINKED ISSUE #45441: OPENSPACE dark cover all and block interaction.
Running /tg/ revision: 2019-07-24
origin/master: 593631c4d5e02b6f5ad76c63e7e00b9be5c804a1
HEAD: f6cd3e011fdfbc01d56ec287cd70571463a6fffd
## Reproduction:
Step on open space and see himself under dark.

### LINKED ISSUE #47072: Objects placed on a wall will show up on the floor besides openspace one Z level above
## Reproduction:
1. Place air alarm
2. Make openspace above the tile the air alarm is 'placed on'
3. Go up one Z level
4. Go stand besides the tile you just placed.
5. Air alarm shows up on the floor

image

### LINKED ISSUE #47102: Rocks on the Z level below will spill over onto the one above
Reproduction:
1. Make a multi Z map
2. Make a lower Z level contain rocks
3. Place openspace above the rocks one Z level above
4. It will now spill over

image

### LINKED ISSUE #47164: Can't build on overspace that show overspace.
Title.

## Reproduction:

Go to 3 level of multiz debug map and try to build over deep chasm that show you 1 level.

### PR DESCRIPTION
**Pull request by Dennok**
## About The Pull Request

fix #47164
fix #45441

fix #47102 fix #47072 While vis flags used and transfer obj, turf and mob icons to openspace plan that i place below all turfs.

## Why It's Good For The Game

MultiZ Cool

## Changelog
:cl:
fix: on 513 Openspace dark no more covers all on openspace turf.
fix: on 513 You now can build on openspace that above openspace.
fix: on 513 While vis flags used and transfer obj icon to openspace plan that placed below all turfs, any things that out of turf bounds now shown correctly below.
/:cl:

openspace 1

scrnshot1

In first try i make possible examine of items below by alt click, but this make problems when i try make floor over fulltile window.

Known issue, in 2 level pit (when you stay on 3 lvl of station and see to pit above spawn room), things from 1 lvl and 2 lvl shown in one pile.
Now i think, this can be fixed by make dynamical openspace plane to make lower levels of openspace lower plane. But dont see clear way to implement it.

But while we don't have 3 lvl maps (except for debug) this don't big issue.

### CODE REVIEW

**Comment by MrPerson** [comment]
Because this sets vis_flags, this PR will break 512 support. Not too big a deal, just something to be aware of.

**Comment by actioninja** [comment]
We aren't running on 512 anymore so I think that's safe to do.

**Comment by PKPenguin321** [comment]
I think there's a bounty on these bugs if you're not aware, check the coding board on the forum

**Comment by optimumtact** [comment]
we can break 512 support, but I wish we had a better way to announce that somehow

**Comment by Dennok** [comment]
> we can break 512 support, but I wish we had a better way to announce that somehow

I can set 512 support like in here https://github.com/tgstation/tgstation/blob/73dc129605019869ff08687cae4c41b5a1eb2050/code/modules/power/solar.dm#L26-L38

but this can disable proper item layering.

**Comment by MMMiracles** [comment]
dennok you are my hero

**Comment by AnturK** [comment]
Not a fan of setting these on base types, will cause headscratching for anything using vis contents for anything else.

**Comment by Dennok** [comment]
> Not a fan of setting these on base types, will cause headscratching for anything using vis contents for anything else.

Problem that we dont have inverted vis_flags to things that have something in vis_contents to control way of displaying.

File: code/game/turfs/open/openspace.dm [hunk 1]
```diff
@@ -14,8 +25,13 @@
 
 /turf/open/openspace/Initialize() // handle plane and layer here so that they don't cover other obs/turfs in Dream Maker
 	. = ..()
-	plane = FLOOR_OPENSPACE_PLANE
 	layer = OPENSPACE_LAYER
+	plane = OPENSPACE_PLANE
+
+#if DM_VERSION >= 513
+	vis_contents += new /atom/movable/openspace_backdrop()
```
Comment (Dennok) [inline code/game/turfs/open/openspace.dm]: I think it should be done in another way, but I don't see it now.
Earlier, openspace turf played this role, but while objects from below inherit the openspace plane, we need something else that will play the role of darkness of depth.

One way to improve this is to create one openspace_backdrop for all open spaces. But what is the best way to do this, and will it even work?

File: code/game/turfs/open/openspace.dm [hunk 2]
```diff
@@ -14,8 +25,11 @@
 
 /turf/open/openspace/Initialize() // handle plane and layer here so that they don't cover other obs/turfs in Dream Maker
 	. = ..()
-	plane = FLOOR_OPENSPACE_PLANE
+	plane = OPENSPACE_PLANE
 	layer = OPENSPACE_LAYER
+
+	vis_contents += new /atom/movable/openspace_backdrop() //Special grey square for projecting backdrop darkness filter on it.
```
Comment (Dennok) [inline code/game/turfs/open/openspace.dm]: I think it should be done in another way, but I don't see it now.
Earlier, openspace turf played this role, but while objects from below inherit the openspace plane, we need something else that will play the role of darkness of depth.

One way to improve this is to create one openspace_backdrop for all open spaces. But what is the best way to do this, and will it even work?

**Comment by Dennok** [comment]
I think it ready for now.
If you have any issues, write it here or give the link of it in issues.

File: code/game/turfs/open/openspace.dm (hunk 2 above)
Comment (AnturK) [reply code/game/turfs/open/openspace.dm]: Any way this happens this shouldn't be one movable per turf, just a single global/static.

Comment (Dennok) [reply code/game/turfs/open/openspace.dm]: I make it right?

**Pull request merged**